src/featbin/generate-random-cmn-offsets.cc

```diff
@@ -0,0 +1,222 @@
+      cmn_wspecifier = po.GetArg(3);
+    DoubleMatrixWriter cmn_writer(cmn_wspecifier); 
+   
+    srand(srand_seed);
+    int32 num_done = 0, num_err = 0, feat_dim = 0;
+
+    SequentialTokenVectorReader spk2utt_reader(spk2utt_rspecifier);
+    RandomAccessBaseFloatMatrixReader feat_reader(feat_rspecifier); 
+
+    RandomAccessBaseFloatVectorReader weights_reader(weights_rspecifier);
+
+    Matrix<double> global_stats,
+      spk_cov;
+    bool is_global_init = false;
+    int32 num_spks = 0;
+    Vector<double> spk_mean_sum;
+    std::map<std::string, Vector<double> > spk_means;
+    std::map<std::string, std::vector<std::string> > spk_uttlist;
+    for (; !spk2utt_reader.Done(); spk2utt_reader.Next()) {
+      std::string spk = spk2utt_reader.Key(); 
+      const std::vector<std::string> &uttlist = spk2utt_reader.Value();
+      spk_uttlist[spk] = uttlist;
+      bool is_init = false;
+      Matrix<double> stats;
+      for (size_t i = 0; i < uttlist.size(); i++) {
+        std::string utt = uttlist[i];
+        if (!feat_reader.HasKey(utt)) {
+            KALDI_WARN << "Did not find features for utterance " << utt;
+            num_err++;
+            continue;
+        }
+        const Matrix<BaseFloat> &feats = feat_reader.Value(utt);
+        // Initialize local stat to accumulate stats per speaker
+        if (!is_init) {
+          InitCmvnStats(feats.NumCols(), &stats);
+          feat_dim = feats.NumCols();
+          is_init = true;
+        }
+
+       
+        if (!AccCmvnStatsWrapper(utt, feats, &weights_reader, &stats)) 
+          num_err++;
+        else
+          num_done++;
+      }
+      Vector<double> spk_mean(stats.Row(0).Range(0, feat_dim));
+      double count = stats(0, feat_dim);
+      spk_mean.Scale(1.0/count);
+      spk_means[spk] = spk_mean;
+      // Initialize global stats to accumulate stats for all data.
+      if (!is_global_init) {
+        InitCmvnStats(feat_dim, &global_stats);
+        spk_cov.Resize(feat_dim, feat_dim);
+        spk_mean_sum.Resize(feat_dim);
+        is_global_init = true;
+      }
+      
+      // spk_cov is the centered covariance of the speaker means
+      // as sum_{i=0}^n (m_i - g)^T (m_i - g), where m_i is spk mean for 
+      // speaker i and g is the gloabl mean.
+      // add sum_{i=0}^n m_i^T * m_i
+      num_spks++;
+      spk_cov.AddVecVec(1.0, spk_mean, spk_mean);
+      spk_mean_sum.AddVec(1.0, spk_mean);
+      global_stats.AddMat(1.0, stats);
+    }
+    // global_mean (g) is the global mean for all frames.
+    Vector<double> global_mean(global_stats.Row(0).Range(0, feat_dim));
+
+    double global_count = global_stats(0, feat_dim);
+    global_mean.Scale(1.0 / global_count);
+
+    Vector<double> global_mean2(spk_mean_sum); 
+    global_mean2.Scale(1.0 / num_spks);
+    // add term -2 * (sum_{i=0}^n m_i^T) g
+    spk_cov.AddVecVec(-2.0, spk_mean_sum, global_mean2);
+    
+    // add term gT * g
+    spk_cov.AddVecVec(1.0, global_mean2, global_mean2);
```

Comment:
... also, I think it would be best if you add some 'self-testing' code to this program.  E.g. take the covariance of the speakers after adding the CMN offsets, multiply it on each side by the inverse-stddev of the original speaker covariance, print out the eigenvalues, and make sure they are close to 1 (if it's been done in such a way that the speaker covariance is supposed to be the same).